### PR TITLE
SUBMARINE-964. Pull notebook image is very time consuming

### PR DESCRIPTION
**Pull request by FatalLin**
### What is this PR for?
Pulling the image of notebook is a quite time consuming process, and it will make user confuse that what the development is waiting for, so we add a new status - Pulling to indicate  the pending is caused by pulling operation.

### What type of PR is it?
Improvement.

### Todos
N/A

### What is the Jira issue?
https://issues.apache.org/jira/browse/SUBMARINE-964
### How should this be tested?
Observe the status in notebook page after creating a notebook.
### Screenshots (if appropriate)
截圖 2021-09-01 下午6 44 52

### Questions:
* Do the license files need updating? No
* Are there breaking changes for older versions? No
* Does this need new documentation? No

### CODE REVIEW

**Review by pingsutw — APPROVED** [review]
+1, awesome, thanks.

**Review by lowc1012 — COMMENTED** [review]
@FatalLin  Thanks!
Would it be better to improve below method?
org.apache.submarine.server.submitter.k8s.util.NotebookUtils#processStatus

**Comment by FatalLin** [comment]
@lowc1012 oops I didn't notice the method before, I'll try try the split the codes into NotebookUtils and K8sSubmitter, maybe you could come back and check later.

**Comment by FatalLin** [comment]
> @lowc1012 oops I didn't notice the method before, I'll try try the split the codes into NotebookUtils and K8sSubmitter, maybe you could come back and check later.

@lowc1012 I just checked

**Comment by FatalLin** [comment]
@lowc1012 I just checked the current flow and found out it would be pretty complicate if we move the codes into NotebookUtils.

current flow:
`NotebookManager -> k8sSubmitter -> NotebookUtil`
if we want to move the codes into NotebookUtil, notebookUtil will need to call k8sSubmitter to get further details of pods and event which will cause a calling loop, and I'm don't think it's a good design for us.
I had an alternative solution is moving part of the codes into NotebookManager, like the notebookManager will check is pod waiting due to pulling operation(and this is also a specific case for k8sSubmitter, even yarn is no longer supported).
How do you think about these?

**Comment by lowc1012** [comment]
@FatalLin
As we discussed offline, we can raise another issue to refactor K8sSubmitter and NotebookManager.
Thanks for your suggestion!

**Review by pingsutw — APPROVED** [review]

**Comment by FatalLin** [comment]
@kevin85421 @pingsutw @lowc1012  Hi guys, thanks for the discussion in the last two days!
I just pushed the last commit of this PR, and this should be the final version of it.
per discussion with Kai-Hsun, we added a new Map property into Notebook Meta Class, which is also could be leveraged later if we allow user to enter their own label from front-end.
We put the notebookId into the label properties, and all the operation depends on the name has been modified, they are depending on the notebookId now.
I also added the check while notebook creating operation to make sure the notebook name is unique.

File: submarine-server/server-submitter/submitter-k8s/src/main/java/org/apache/submarine/server/submitter/k8s/K8sSubmitter.java [hunk 1]
```diff
@@ -429,10 +429,30 @@ public Notebook findNotebook(NotebookSpec spec) throws SubmarineRuntimeException
...
+        String podLabelSelector = String.format("%s=%s", NotebookCR.NOTEBOOK_ID,
+            spec.getMeta().getLabels().get(NotebookCR.NOTEBOOK_ID).toString());
+
+        V1PodList podList = coreApi.listNamespacedPod(namespace, null, null, null, podLabelSelector,
+            null, null, null, null);
+        String podName = podList.getItems().get(0).getMetadata().getName();
+        String fieldSelector = String.format("involvedObject.name=%s", podName);
+        V1EventList events = coreApi.listNamespacedEvent(namespace, null, null, fieldSelector,
+            null, null, null, null, null);
+        V1Event lastestEvent = events.getItems().get(events.getItems().size() - 1);
```
Comment (lowc1012) [inline submarine-server/server-submitter/submitter-k8s/src/main/java/org/apache/submarine/server/submitter/k8s/K8sSubmitter.java:449]: latestEvent?

File: submarine-server/server-core/src/main/java/org/apache/submarine/server/notebook/NotebookManager.java [hunk 2]
```diff
@@ -81,7 +83,17 @@ public Notebook createNotebook(NotebookSpec spec) throws SubmarineRuntimeExcepti
...
     spec.getMeta().setName(lowerName);
     NotebookId notebookId = generateNotebookId();
+
+    Map<String, String> labels = spec.getMeta().getLabels();
+
+    if (labels == null) {
+      labels = new HashMap<>();
+    }
+    labels.put("notebook-owner-id", spec.getMeta().getOwnerId());
+    labels.put("notebookId", notebookId.toString());
```
Comment (lowc1012) [inline submarine-server/server-core/src/main/java/org/apache/submarine/server/notebook/NotebookManager.java:93]: Use "notebook-id" ?
ref: [syntax-and-character-set](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set)

**Comment by FatalLin** [comment]
@lowc1012 good call! both suggestion had been integrated into this PR.

**Review by pingsutw — APPROVED** [review]
Thank you @FatalLin @lowc1012.

**Review by lowc1012 — APPROVED** [review]
+1, Thanks @FatalLin
Will merge if no more comments.

**Pull request closed without merging**